src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,685 @@
+    abstract public function folded(): self;
+
+    /**
+     * @return static
+     */
+    public function ignoreCase(): self
+    {
+        $str = clone $this;
+        $str->ignoreCase = true;
+
+        return $str;
+    }
+
+    /**
+     * @param string|string[] $needle
+     */
+    public function indexOf($needle, int $offset = 0): ?int
+    {
+        if (!\is_array($needle) && !$needle instanceof \Traversable) {
+            throw new \TypeError(sprintf('Method "%s()" must be overridden by class "%s" to deal with non-iterable values.', __FUNCTION__, \get_class($this)));
+        }
+
+        $i = \PHP_INT_MAX;
+
+        foreach ($needle as $n) {
+            $j = $this->indexOf((string) $n, $offset);
+
+            if (null !== $j && $j < $i) {
+                $i = $j;
+            }
+        }
+
+        return \PHP_INT_MAX === $i ? null : $i;
+    }
+
+    /**
+     * @param string|string[] $needle
+     */
+    public function indexOfLast($needle, int $offset = 0): ?int
+    {
+        if (!\is_array($needle) && !$needle instanceof \Traversable) {
+            throw new \TypeError(sprintf('Method "%s()" must be overridden by class "%s" to deal with non-iterable values.', __FUNCTION__, \get_class($this)));
+        }
+
+        $i = null;
+
+        foreach ($needle as $n) {
+            $j = $this->indexOfLast((string) $n, $offset);
+
+            if (null !== $j && $j > $i) {
+                $i = $offset = $j;
+            }
+        }
+
+        return $i;
+    }
+
+    public function isEmpty(): bool
+    {
+        return '' === $this->string;
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function join(array $strings): self;
+
+    public function jsonSerialize(): string
+    {
+        return $this->string;
+    }
+
+    abstract public function length(): int;
+
+    /**
+     * @return static
+     */
+    abstract public function lower(): self;
+
+    /**
+     * Matches the string using a regular expression.
+     *
+     * Pass PREG_PATTERN_ORDER or PREG_SET_ORDER as $flags to get all occurrences matching the pattern.
+     *
+     * @return array All matches in a multi-dimensional array ordered according to flags
+     */
+    abstract public function match(string $pattern, int $flags = 0, int $offset = 0): array;
+
+    /**
+     * @return static
+     */
+    abstract public function padBoth(int $length, string $padStr = ' '): self;
+
+    /**
+     * @return static
+     */
+    abstract public function padEnd(int $length, string $padStr = ' '): self;
+
+    /**
+     * @return static
+     */
+    abstract public function padStart(int $length, string $padStr = ' '): self;
+
+    /**
+     * @return static
+     */
+    abstract public function prepend(string ...$prefix): self;
+
+    /**
+     * @return static
+     */
+    public function repeat(int $multiplier): self
+    {
+        $str = clone $this;
+        $str->string = str_repeat($str->string, 0 > $multiplier ? 0 : $multiplier);
```

Comment:
What about throwing instead if `$multiplier` is less than `0`?

Reply:
I'm not sure it would be more useful than the current behavior. I favored not throwing when we don't have to.

Reply:
Throwing would allow detecting mistakes when you expected a positive number to be passed which was somehow inverted to a negative number. Having this silently ignored will unnecessarily make debugging harder IMO.

Reply:
OK, updated.